File: mbfpp/code/ClojureExamples/src/mbfpp/functional/fb/composition_examples.clj
(ns mbfpp.functional.fb.composition-examples)

(defn append-a [s] (str s "a"))
(defn append-b [s] (str s "b"))
(defn append-c [s] (str s "c"))

(def append-cba (comp append-a append-b append-c))

(def request 
  {:headers 
   {"Authorization" "auth"
    "X-RequestFingerprint" "fingerprint"}
   :body "body"})

(defn check-authorization [request]
  (let [auth-header (get-in request [:headers "Authorization"])]
    (assoc
      request
      :principal
      (if-not (nil? auth-header)
        "AUser"))))

(defn log-fingerprint [request]
  (let [fingerprint (get-in request [:headers "X-RequestFingerprint"])]
    (println (str "FINGERPRINT=" fingerprint))
    request))

(defn compose-filters [filters]
  (reduce 
    (fn [all-filters, current-filter] (comp all-filters current-filter))
    filters))

File: mbfpp/code/ScalaExamples/src/main/scala/com/mblinn/mbfpp/functional/fb/CompositionExamples.scala
package com.mblinn.mbfpp.functional.fb

object ScalaExamples {

  val appendA = (s: String) => s + "a"
  val appendB = (s: String) => s + "b"
  val appendC = (s: String) => s + "c"
  
  val appendCBA = appendA compose appendB compose appendC
  
  case class HttpRequest(
      headers: Map[String, String], 
      payload: String, 
      principal: Option[String] = None)
  
  def checkAuthorization(request: HttpRequest) = {
    val authHeader = request.headers.get("Authorization")
    val mockPrincipal = authHeader match {
      case Some(headerValue) => Some("AUser")
      case _ => None
    }
    request.copy(principal = mockPrincipal)
  }
  def logFingerprint(request: HttpRequest) = {
    val fingerprint = request.headers.getOrElse("X-RequestFingerprint", "")
    println("FINGERPRINT=" + fingerprint)
    request
  }
  
  def composeFilters(filters: Seq[Function1[HttpRequest, HttpRequest]]) =
    filters.reduce { 
      (allFilters, currentFilter) => allFilters compose currentFilter
    }

}